File: cpp/0093.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;
void show_vv(vector<vector<int>> pos) {
  for (int i = 0; i < pos.size(); i++) {
    for (int j = 0; j < pos[i].size(); j++) {
      cout << pos[i][j] << ".";
    }
    cout << endl;
  }
}
class Solution {
public:
  string vector_int_to_string(vector<int> a) {
    if (a.size() == 0)
      return string("");
    string s(to_string(a[0]));
    for (int i = 1; i < a.size(); i++) {
      s.append(".");
      s.append(to_string(a[i]));
    }
    return s;
  }
  void show_vec(vector<int> now) {
    cout << "before show vec" << endl;
    cout << vector_int_to_string(now) << endl;
  }
  void go(vector<string> &a, vector<int> now, vector<vector<int>> pos, int p,
          string s) {
    if (now.size() > 4)
      return;
    // cout << vector_int_to_string(now) << " " << p << endl;
    if (p == s.length()) {
      if (now.size() == 4) {
        string ip = vector_int_to_string(now);
        // cout << "found" << ip << endl;
        a.push_back(ip);
      }
      return;
    }
    vector<int> canbe = pos[p];
    for (int i = 0; i < canbe.size(); i++) {
      int end = canbe[i];
      int value = stoi(s.substr(p, end - p));
      now.push_back(value);
      go(a, now, pos, end, s);
      now.pop_back();
    }
  }
  vector<string> restoreIpAddresses(string s) {
    vector<string> a;
    vector<vector<int>> pos;
    for (int i = 0; i < s.length(); i++) {
      vector<int> p;
      pos.push_back(p);
    }
    for (int i = 0; i < s.length(); i++) {
      if (s[i] == '0') {
        pos[i].push_back(i + 1);
        continue;
      }
      for (int j = i + 1; j <= s.length(); j++) {
        string sub = s.substr(i, j - i);
        if (stoi(sub) < 256) {
          pos[i].push_back(j);
        } else {
          break;
        }
      }
    }
    // show_vv(pos);
    // cout << "==" << endl;
    vector<int> now;
    go(a, now, pos, 0, s);
    return a;
  }
};

File: rust/src/bin/0093.rs
struct Solution;

fn go(ans: &mut Vec<String>, ind: usize, s: &Vec<i32>, now: &mut Vec<i32>) {
    if ind == s.len() {
        if now.len() != 4 {
            return;
        }
        let mut n = String::new();
        let mut is_first = true;
        for i in now {
            if is_first {
                is_first = false;
            } else {
                n.push('.');
            }
            n.push_str(i.to_string().as_str());
        }
        ans.push(n);
        return;
    }
    if now.len() >= 4 {
        return;
    }
    let mut ip = 0;
    for i in 0..3 {
        if ind + i >= s.len() {
            break;
        }
        ip = ip * 10 + s[ind + i];
        if ip > 255 {
            break;
        }
        now.push(ip);
        go(ans, ind + i + 1, s, now);
        now.pop();
        if ip == 0 && i == 0 {
            //不可能以0开头
            break;
        }
    }
}

impl Solution {
    pub fn restore_ip_addresses(s: String) -> Vec<String> {
        let mut ans = vec!();
        let mut ss = vec!();
        for i in s.chars() {
            ss.push(i as i32 - '0' as i32);
        }
        println!("{:?}", ss);
        let mut now = vec!();
        go(&mut ans, 0, &ss, &mut now);
        ans
    }
}

fn main() {
    let ans = Solution::restore_ip_addresses("1111".to_string());
    println!("{:?}", ans);
}